pkg/hive/hive.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package hive
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/addressbook"
+	"github.com/ethersphere/bee/pkg/discovery"
+	"github.com/ethersphere/bee/pkg/hive/pb"
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/p2p"
+	"github.com/ethersphere/bee/pkg/p2p/protobuf"
+	"github.com/ethersphere/bee/pkg/swarm"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+const (
+	protocolName    = "hive"
+	protocolVersion = "1.0.0"
+	peersStreamName = "peers"
+	messageTimeout  = 5 * time.Second // maximum allowed time for a message to be read or written.
+	maxBatchSize    = 50
+)
+
+type Service struct {
+	streamer    p2p.Streamer
+	addressBook addressbook.GetterPutter
+	logger      logging.Logger
+}
+
+type Options struct {
+	Streamer    p2p.Streamer
+	AddressBook addressbook.GetterPutter
+	Logger      logging.Logger
+}
+
+func New(o Options) *Service {
+	return &Service{
+		streamer:    o.Streamer,
+		logger:      o.Logger,
+		addressBook: o.AddressBook,
+	}
+}
+
+func (s *Service) Protocol() p2p.ProtocolSpec {
+	return p2p.ProtocolSpec{
+		Name:    protocolName,
+		Version: protocolVersion,
+		StreamSpecs: []p2p.StreamSpec{
+			{
+				Name:    peersStreamName,
+				Handler: s.peersHandler,
+			},
+		},
+	}
+}
+
+func (s *Service) BroadcastPeers(ctx context.Context, addressee swarm.Address, peers ...discovery.BroadcastRecord) error {
+	for len(peers) > maxBatchSize {
```

Comment:
can be shortened to one for loop:
```go
        max := maxBatchSize
	for len(peers) > 0 {
                if max > len(peers) { max = len(peers) }
		if err := s.sendPeers(ctx, addressee, peers[:max]); err != nil {
			return err
		}

		peers = peers[max:]
	}
```

Reply:
I agree, nice. It does have an additional if statement in the for loop and an additional variable, but it does look more compact. I like both solutions, so I can change if you like this one more?

Reply:
Changed for the shorter loop.